src/core/raster/qgspalettedrasterrenderer.cpp

```diff
@@ -472,37 +475,98 @@ QgsPalettedRasterRenderer::ClassData QgsPalettedRasterRenderer::classDataFromRas
   if ( !raster )
     return ClassData();
 
-  // get min and max value from raster
-  QgsRasterBandStats stats = raster->bandStatistics( bandNumber, QgsRasterBandStats::Min | QgsRasterBandStats::Max, QgsRectangle(), 0, feedback );
-  if ( feedback && feedback->isCanceled() )
-    return ClassData();
-
-  double min = stats.minimumValue;
-  double max = stats.maximumValue;
-  // need count of every individual value
-  int bins = std::ceil( max - min ) + 1;
-  if ( bins <= 0 )
-    return ClassData();
-
-  QgsRasterHistogram histogram = raster->histogram( bandNumber, bins, min, max, QgsRectangle(), 0, false, feedback );
-  if ( feedback && feedback->isCanceled() )
-    return ClassData();
-
-  double interval = ( histogram.maximum - histogram.minimum + 1 ) / histogram.binCount;
-
   ClassData data;
-
-  double currentValue = histogram.minimum;
   double presentValues = 0;
-  for ( int idx = 0; idx < histogram.binCount; ++idx )
+
+  // Collect unique values for float rasters
+  if ( raster->dataType( bandNumber ) == Qgis::DataType::Float32 || raster->dataType( bandNumber ) == Qgis::DataType::Float64 )
   {
-    int count = histogram.histogramVector.at( idx );
-    if ( count > 0 )
+    const QgsRectangle fullExtent { raster->extent() };
+    QList<double> values;
```

Comment:
A list isn't a great choice here -- could you use an ordered set instead? or unordered set and sort after collecting the values?